libyul/backends/wasm/WordSizeTransform.cpp

```diff
@@ -0,0 +1,238 @@
+				if (
+					!varDecl.value ||
+					varDecl.value->type() == typeid(FunctionalInstruction) ||
+					varDecl.value->type() == typeid(FunctionCall)
+				)
+				{
+					if (varDecl.value) visit(*varDecl.value);
+					rewriteVarDeclList(varDecl.variables);
+					return boost::none;
+				}
+				else if (
+					varDecl.value->type() == typeid(Identifier) ||
+					varDecl.value->type() == typeid(Literal)
+				)
+				{
+					yulAssert(varDecl.variables.size() == 1, "");
+					auto newRhs = expandValue(*varDecl.value);
+					auto newLhs = generateU64IdentifierNames(varDecl.variables[0].name);
+					vector<Statement> ret;
+					for (int i = 0; i < 4; i++)
+						ret.push_back(
+							VariableDeclaration{
+								varDecl.location,
+								{TypedName{varDecl.location, newLhs[i], "u64"_yulstring}},
+								std::move(newRhs[i])
+							}
+						);
+					return ret;
+				}
+				else
+					yulAssert(false, "");
+			}
+			else if (_s.type() == typeid(Assignment))
+			{
+				Assignment& assignment = boost::get<Assignment>(_s);
+				yulAssert(assignment.value, "");
+				if (
+					assignment.value->type() == typeid(FunctionalInstruction) ||
+					assignment.value->type() == typeid(FunctionCall)
+				)
+				{
+					if (assignment.value) visit(*assignment.value);
+					rewriteIdentifierList(assignment.variableNames);
+					return boost::none;
+				}
+				else if (
+					assignment.value->type() == typeid(Identifier) ||
+					assignment.value->type() == typeid(Literal)
+				)
+				{
+					yulAssert(assignment.variableNames.size() == 1, "");
+					auto newRhs = expandValue(*assignment.value);
+					YulString lhsName = assignment.variableNames[0].name;
+					vector<Statement> ret;
+					for (int i = 0; i < 4; i++)
+						ret.push_back(
+							Assignment{
+								assignment.location,
+								{Identifier{assignment.location, m_variableMapping.at(lhsName)[i]}},
+								std::move(newRhs[i])
+							}
+						);
+					return ret;
+				}
+				else
+					yulAssert(false, "");
+			}
+			else
+				visit(_s);
+			return boost::none;
+		}
+	);
+}
+
+array<YulString, 4> WordSizeTransform::generateU64IdentifierNames(YulString const& _s)
+{
+	yulAssert(m_variableMapping.find(_s) == m_variableMapping.end(), "");
+	array<YulString, 4> ret;
+	for (int i = 0; i < 4; i++)
+	{
+		auto newName = m_nameDispenser.newName(_s);
+		m_variableMapping[_s][i] = newName;
+		ret[i] = newName;
+	}
+	return ret;
+}
+
+array<unique_ptr<Expression>, 4> WordSizeTransform::expandValue(Expression const& _e)
+{
+	array<unique_ptr<Expression>, 4> ret;
+	if (_e.type() == typeid(Identifier))
+	{
+		Identifier const& id = boost::get<Identifier>(_e);
+		for (int i = 0; i < 4; i++)
+			ret[i] = make_unique<Expression>(Identifier{id.location, m_variableMapping.at(id.name)[i]});
+	}
+	else if (_e.type() == typeid(Literal))
+	{
+		Literal const& lit = boost::get<Literal>(_e);
+		yulAssert(lit.kind == LiteralKind::Number, "Non-numeric literal unimplemented.");
```

Comment:
yul::ValueOfLiteral?

Reply:
ok, also allows other kinds of Literal. They are all u256 anyway.

Reply:
(allows the other kinds of Literal by removing this assertion)